MM-45853 : Add a confirmation to the mark all as read button in Threads

kyeongsoosoo opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->

- Add `mark_all_threads_as_read_modal`
- Fix related e2e test

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->

Fixes https://github.com/mattermost/mattermost-server/issues/21142
Jira: https://mattermost.atlassian.net/browse/MM-45853

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs.

For an easier comparison of UI changes a table (template below) can be used.

|  before  |  after  |
|----|----|
| <insert before screenshot here> | <insert after screenshot here> |

-->


https://user-images.githubusercontent.com/59413880/193645629-f7cb27b3-c4b2-4660-aec6-6b7b16bf2c86.mov



#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note
Added confirmation modal before marking all thread as read.
```

mattermod commented:
E2E tests not automatically triggered, because PR has no approval yet. Please ask a developer to review and then try again to attach the QA label.

michelengelen commented on components/threading/global_threads/mark_all_threads_as_read_modal/mark_all_threads_as_read_modal.tsx:
```diff
@@ -0,0 +1,55 @@
+        >
+            <div className='mark_all_threads_as_read_modal__body'>
+                <span>
+                    { formatMessage({
```
nit

```suggestion
                    {formatMessage({
```

kyeongsoosoo commented:
@michelengelen Thanks for the review.
+) I added  `onExited` prop.

michelengelen reviewed (approved):
LGTM

matthewbirtch reviewed (changes requested):
Seeing one small issue on mobile webview - there seems to be a horizontal gradient between the title and the body text. This seems to happen in browser widths <= 640px
<img width="240" alt="image" src="https://user-images.githubusercontent.com/2040554/194412696-e1b4e2b1-1954-4438-b3a5-5a10b4255150.png">

Other than that, this is great!

kyeongsoosoo commented:
@matthewbirtch Thanks. I fixed it!
<img width="697" alt="스크린샷 2022-10-07 오전 10 51 09" src="https://user-images.githubusercontent.com/59413880/194449526-aa0c6f2d-7d8a-4a17-a74c-b78e4766f0c9.png">

matthewbirtch reviewed (approved):
Thanks @kyeongsoosoo! Looks good to me.

michelengelen commented:
@kyeongsoosoo could you merge `master` into your branch and resolve the conflicts please? If you need advice/help feel free to ping me

jgilliam17 commented:
Thanks @kyeongsoosoo for fixing the e2es.
Can you also expand the cypress test to check if canceling modal is working as expected?
You can see the suggested steps here https://automation-test-cases.vercel.app/test/MM-T4448 Step 2 is new and not automated yet.
If not, I can file as a help wanted follow up. Let me know - Thanks again 🙂

kyeongsoosoo commented:
@jgilliam17  I added the e2e test.
I follow the suggested steps, and I found some errors.

> 	6.	Click on “Mark all as unread” button again
      1. Verify mark_all_threads_as_read_modal is open
      2. Verify thread is marked as read, Unread tab no longer shows blue dot, right pane shows “Looks like you’re all caught up”

Thread should not be marked as read when Mark all as unread” button is clicked, so I added steps without checking it.

jgilliam17 commented:
Thanks for catching that @kyeongsoosoo   that’s totally my fault for copying and pasting steps without confirming “verify…” part. I’ll update the step is Zephyr to reflect the corrections you’ve made.

M-ZubairAhmed commented on components/threading/global_threads/mark_all_threads_as_read_modal/mark_all_threads_as_read_modal.tsx:
```diff
@@ -0,0 +1,55 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+import * as React from 'react';
+
```
nit : no line space in between

M-ZubairAhmed commented on components/threading/global_threads/thread_list/thread_list.tsx:
```diff
@@ -32,6 +32,10 @@ import CRTUnreadTutorialTip
 
 import {getIsMobileView} from 'selectors/views/browser';
 
+import MarkAllThreadsAsReadModal, {MarkAllThreadsAsReadModalProps} from '../mark_all_threads_as_read_modal';
+
+import {closeModal, openModal} from 'actions/views/modals';
+
```
nit
```suggestion
import {closeModal, openModal} from 'actions/views/modals';

import MarkAllThreadsAsReadModal, {MarkAllThreadsAsReadModalProps} from '../mark_all_threads_as_read_modal';
```

M-ZubairAhmed commented on components/threading/global_threads/thread_list/thread_list.tsx:
```diff
@@ -146,6 +150,10 @@ const ThreadList = ({
         return {data: true};
     }, [currentTeamId, data, unread, selectedThreadId]);
 
+    const handleCloseMarkAllAsReadModal = useCallback(() => {
+        dispatch(closeModal(ModalIdentifiers.MARK_ALL_THREADS_AS_READ));
+    }, []);
+
```
this doesnt have to be inside the useCallback, since this function is not a prop being passed down to a component https://reactjs.org/docs/hooks-reference.html#usecallback

kyeongsoosoo replied:
Thanks for pointing it out :)
I moved this code into `handleOpenMarkAllAsReadModal`
```ts
const handleOpenMarkAllAsReadModal = useCallback(() => {
    const handleCloseMarkAllAsReadModal = () => {
        dispatch(closeModal(ModalIdentifiers.MARK_ALL_THREADS_AS_READ));
    };

    const handleConfirm = () => {
        handleAllMarkedRead();
        handleCloseMarkAllAsReadModal();
    };
    ...
```

M-ZubairAhmed reviewed (changes requested):
Code looks awesome and very understandable. Thank you. 
Just a few nits

jgilliam17 commented on e2e/cypress/tests/integration/collapsed_reply_threads/global_threads_spec.js:
```diff
@@ -320,4 +329,88 @@ describe('Collapsed Reply Threads', () => {
         });
     });
+
+    it('MM-T4448 CRT - L16 - Cancel “Mark all as unread” modal', () => {
```
```suggestion
    it('MM-T4448_2 CRT - L16 - Cancel “Mark all as unread” modal', () => {
```

Also please update line 267 with updated test case id `it('MM-T4448_1 CRT - L16 - Use “Mark...`

kyeongsoosoo replied:
@jgilliam17 I updated it!

jgilliam17 commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/249652

jgilliam17 reviewed (commented):
Thanks @kyeongsoosoo 
Verified confirmation dialog for "mark all as read" button
Please update the test case id as suggested below.
I'm just waiting for the cypress cycle to finish before approving

jgilliam17 reviewed (approved):
Thanks for the update @kyeongsoosoo 
[E2E report](https://mattermost-cypress-report.s3.amazonaws.com/249652-b2b51b5-onprem-ent-webapp-pr-11260-mm-ee-test:b2b51b5/mochawesome.html) - no PR related failures - looks good to merge.
